pilz_control/test/unittest_cartesian_speed_monitor.cpp

```diff
@@ -0,0 +1,255 @@
+
+using pilz_control::CartesianSpeedMonitor;
+
+class CartesianSpeedMonitorTest : public testing::Test
+{
+protected:
+  void SetUp() override;
+
+protected:
+  moveit::core::RobotModelPtr model_;
+  std::vector<std::string> joint_names_;
+};
+
+void CartesianSpeedMonitorTest::SetUp()
+{
+  moveit::core::RobotModelBuilder builder("test_robot", "base");
+
+  // keep in mind that we always rotate around the x-axis, so
+  // by the following transformation to the origin of link1 we obtain
+  // a rotation of link1 around the (base) y-axis
+  geometry_msgs::Pose origin1;
+  tf2::Quaternion quat1;
+  quat1.setRPY(0.0, 0.0, -0.5 * M_PI);
+  origin1.orientation = tf2::toMsg(quat1);
+
+  // link2 is rotated about the z-axis tilted about 45 degree in the x-z-plane
+  // the default position is perpendicular to link1 and parallel to the rotation axis of joint1
+  geometry_msgs::Pose origin2;
+  origin2.position.z = 1.0;
+  tf2::Quaternion quat2;
+  quat2.setRPY(0.0, -0.25 * M_PI, 0.5 * M_PI);
+  origin2.orientation = tf2::toMsg(quat2);
+
+  builder.addChain("base->link1->link2", "continuous", { origin1, origin2 });
+
+  geometry_msgs::Pose origin3;
+  origin3.position.y = 1.0;
+
+  // as a result if we fix joint2 such that link2 is in perpendicular position and move joint1,
+  // the maximal velocity is in the origins of link2 and link3 which move equally fast
+  // if we fix joint2 such that the angle between link1 and link2 is above 90 degree and move joint1,
+  // the origin of link3 moves faster than the origin of link2
+  // if we fix joint2 such that the angle between link1 and link2 is below 90 degree and move joint1,
+  // the origin of link3 moves slower than the origin of link2
+  builder.addChain("link2->link3", "fixed", { origin3 });
+
+  model_ = builder.build();
+  joint_names_ = model_->getVariableNames();
+}
+
+TEST_F(CartesianSpeedMonitorTest, testUnmatchedJointCount)
+{
+  joint_names_.resize(joint_names_.size()-1);
+  EXPECT_THROW(CartesianSpeedMonitor monitor(joint_names_, model_), pilz_control::CartesianSpeedMonitorException);
+}
+
+/**
+ * @tests{Monitor_Speed_of_all_links_until_TCP,
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testLinkSpeedNoMovement)
+{
+  robot_state::RobotStatePtr state(new robot_state::RobotState(model_));
+  state->setToDefaultValues();
+  state->updateLinkTransforms();
+  moveit::core::LinkModel* link = model_->getLinkModel("link2");
+  double time_delta = 1.0;
+
+  double speed = CartesianSpeedMonitor::linkSpeed(state, state, link, time_delta);
+  EXPECT_NEAR(speed, 0.0, SPEED_COMPARISON_TOLERANCE);
+
+  link = model_->getLinkModel("link3");
+  speed = CartesianSpeedMonitor::linkSpeed(state, state, link, time_delta);
+  EXPECT_NEAR(speed, 0.0, SPEED_COMPARISON_TOLERANCE);
+}
+
+/**
+ * @tests{Monitor_Speed_of_all_links_until_TCP,
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testLinkSpeedMoveJoint1)
```

Comment:
Refactor test name and test docu, it is not very meaningful because right now I would ask: Why are there two tests (`testLinkSpeedMoveJoint1`and `testLinkSpeedMoveJoint2`), basically doing the same thing? => Explain in more detail your test intention.

Reply:
I refactored the tests